DataFormats/Math/interface/EigenComputations.h

```diff
@@ -0,0 +1,440 @@
+      b[i] = (Atb(i_real) - total) / l_i_i;
+    }
+
+    template <typename MatrixType1, typename MatrixType2, typename MatrixType3>
+    __device__ void solve_forward_subst_matrix(MatrixType1& A,
+                                               MatrixType2 const& pulseMatrixView,
+                                               MatrixType3 const& matrixL) {
+      // FIXME: this assumes pulses are on columns and samples on rows
+      constexpr auto NPULSES = MatrixType2::ColsAtCompileTime;
+      constexpr auto NSAMPLES = MatrixType2::RowsAtCompileTime;
+
+#pragma unroll
+      for (int icol = 0; icol < NPULSES; icol++) {
+        float reg_b[NSAMPLES];
+        float reg_L[NSAMPLES];
+
+// preload a column and load column 0 of cholesky
+#pragma unroll
+        for (int i = 0; i < NSAMPLES; i++) {
+          reg_b[i] = __ldg(&pulseMatrixView.coeffRef(i, icol));
+          reg_L[i] = matrixL(i, 0);
+        }
+
+        // compute x0 and store it
+        auto x_prev = reg_b[0] / reg_L[0];
+        A(0, icol) = x_prev;
+
+// iterate
+#pragma unroll
+        for (int iL = 1; iL < NSAMPLES; iL++) {
+// update accum
+#pragma unroll
+          for (int counter = iL; counter < NSAMPLES; counter++)
+            reg_b[counter] -= x_prev * reg_L[counter];
+
+// load the next column of cholesky
+#pragma unroll
+          for (int counter = iL; counter < NSAMPLES; counter++)
+            reg_L[counter] = matrixL(counter, iL);
+
+          // compute the next x for M(iL, icol)
+          x_prev = reg_b[iL] / reg_L[iL];
+
+          // store the result value
+          A(iL, icol) = x_prev;
+        }
+      }
+    }
+
+    template <typename MatrixType1, typename MatrixType2>
+    __device__ void solve_forward_subst_vector(float reg_b[MatrixType1::RowsAtCompileTime],
+                                               MatrixType1 inputAmplitudesView,
+                                               MatrixType2 matrixL) {
+      constexpr auto NSAMPLES = MatrixType1::RowsAtCompileTime;
+
+      float reg_b_tmp[NSAMPLES];
+      float reg_L[NSAMPLES];
+
+// preload a column and load column 0 of cholesky
+#pragma unroll
+      for (int i = 0; i < NSAMPLES; i++) {
+        reg_b_tmp[i] = inputAmplitudesView(i);
+        reg_L[i] = matrixL(i, 0);
+      }
+
+      // compute x0 and store it
+      auto x_prev = reg_b_tmp[0] / reg_L[0];
+      reg_b[0] = x_prev;
+
+// iterate
+#pragma unroll
+      for (int iL = 1; iL < NSAMPLES; iL++) {
+// update accum
+#pragma unroll
+        for (int counter = iL; counter < NSAMPLES; counter++)
+          reg_b_tmp[counter] -= x_prev * reg_L[counter];
+
+// load the next column of cholesky
+#pragma unroll
+        for (int counter = iL; counter < NSAMPLES; counter++)
+          reg_L[counter] = matrixL(counter, iL);
+
+        // compute the next x for M(iL, icol)
+        x_prev = reg_b_tmp[iL] / reg_L[iL];
+
+        // store the result value
+        reg_b[iL] = x_prev;
+      }
+    }
+
+    /*
+    // TODO: add active bxs
```

Comment:
please remove commented out code, or add a comment why this is needed and when this will be implemented

Reply:
Similar common call  done in 

* ECAL https://github.com/cms-sw/cmssw/pull/31719/files#diff-98f76d25429b5059efd61f6632594e42R17
RecoLocalCalo/EcalRecProducers/plugins/KernelHelpers.h

* HCAL https://github.com/cms-sw/cmssw/pull/31720/files#diff-8ad73af139b7dd05aaecb810d8327c62R625
RecoLocalCalo/HcalRecProducers/src/MahiGPU.cu

would be better here in a common place in DataFormats/Math/interface/EigenComputations.h

the only differences ( originally from the CPU code) are in the 
* nan check (only on ECAL, but would benefit in both)
* convergence parameters (those two integer can be configurable)

       ++iter;
         if (iter % 16 == 0)
           eps_to_use *= 2;

      if (iter % 10 == 0)
           eps_to_use *= 10;

Reply:
Thanks for pointing out the code duplications across the PRs. I was following the links a bit, and I would like to ask for a single, generic implementation of these functions:

- https://github.com/cms-sw/cmssw/pull/31704/files/d07d03072ee00e32717cd2f05b0ce6b76da1bec2#diff-f456782d8083d5d31604451980616cc7728f6be06f43f371e0ef1c67eea2dfdcR282
- https://github.com/cms-sw/cmssw/pull/31720/files#diff-4c3f70988bc0587761bb4c18f9a393f6ef483e3601452ee15e5cd3d34654d1f6R625
- https://github.com/cms-sw/cmssw/pull/31719/files#diff-ed9059ebaf6819819074dce32f5dca13fc1499b457a6a90031adac0b419bcd32R17

This commented-out code block above can then be dropped, and you can reduce the number of lines in the other PRs. Let me know if you will address this, or did I misunderstand the comment?

Reply:
ok, generic implementation of the fnnls, where you want to place ?
not on ECAL, not on HCAL
is fine to have in DataFormats/Math/interface/EigenComputations.h together in the calo::multifit ?
or where you suggest to add all these common fits function ?

Reply:
If this is generic linear algebra, just keep the implementation in DataFormats/Math/interface/EigenComputations.h. I don't want to get stuck on the namespace name here, if there is no better idea in sight and it becomes a blocker for moving forward.

The higher priority item is to remove this code duplication (if possible and feasible) from the other PRs. Thank you for pointing out this particular example, are there other functions which are similarly duplicated? We can of course invest in some automatic checks, but it would be good to hear from the authors.

Reply:
duplicate code removed